### PR TITLE
Fix up sporadically failing tests and ensure tests can run independently

### PR DESCRIPTION
**Pull request by zenspider**
Most of the info is in the commit messages... I should whip up some way to collate them into something I can paste here. 😕

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -12,6 +12,8 @@ before_install:
 - gem update --system
 - gem --version
 - bundle --version
+env:
+  - SLOW=1
```
Comment (miah) [inline .travis.yml:16]: I like this, and that the slow tests get marked as skipped so we don't forget about them =)

**Review by miah — APPROVED** [review]

File: .travis.yml (hunk 1 above)
Comment (zenspider) [reply .travis.yml:16]: Exactly. That was the intent.

If you know of a better way to globally set an env on appveyor, I'd appreciate it.

File: lib/inspec/plugin/v2/loader.rb [hunk 2]
```diff
@@ -113,9 +113,8 @@ def plugin_gem_path
     end
 
     def self.plugin_gem_path
-      # I can't believe there isn't a simpler way of getting this
-      # 2.4.2.p123 => 2.4.0
-      ruby_abi_version = (Gem.ruby_version.segments[0, 2] << 0).join('.')
```
Comment (clintoncwolfe) [inline lib/inspec/plugin/v2/loader.rb:118]: great!

File: lib/matchers/matchers.rb [hunk 3]
```diff
@@ -247,26 +247,26 @@ def try_match(actual, op, expected) # rubocop:disable Metrics/CyclomaticComplexi
     # if actual and expected are strings
     if expected.is_a?(String) && actual.is_a?(String)
       return actual.casecmp(expected) == 0 if op == :==
-      return Gem::Version.new(actual).method(op).call(Gem::Version.new(expected)) if
```
Comment (clintoncwolfe) [inline lib/matchers/matchers.rb:249]: Could you comment as to why these changes are needed to get the tests running more cleanly?

File: lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb [hunk 4]
```diff
@@ -569,6 +569,8 @@ def test_refuse_install_when_plugin_on_exclusion_list
   end
 
   def test_error_install_with_debug_enabled
+    skip "this test requires bundler to pass" unless defined? ::Bundler
```
Comment (clintoncwolfe) [inline lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb:572]: Are there test conditions under which Bundler is not available? We require it to run the test suite...

File: lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb [hunk 5]
```diff
@@ -581,11 +583,13 @@ def test_error_install_with_debug_enabled
 #-----------------------------------------------------------------------------------------#
 #                               inspec plugin update
 #-----------------------------------------------------------------------------------------#
-class PluginManagerCliUpdate < MiniTest::Test
+class PluginManagerCliUpdate < Minitest::Test
   include CorePluginFunctionalHelper
   include PluginManagerHelpers
 
   def test_when_a_plugin_can_be_updated
+    skip "this test requires bundler to pass" unless defined? ::Bundler
```
Comment (clintoncwolfe) [inline lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb:591]: Likewise here, we should have bundler...

File: test/integration/default/controls/audit_spec.rb [hunk 6]
```diff
@@ -14,7 +14,8 @@
 
 # Grab bundled inspec version. This should be the same as the one
 # passed for audit cookbook. If its not, you should do a `bundle install`
-inspec_version = Gem.loaded_specs['inspec'].version.to_s
+inspec_version = Inspec::VERSION
+# or: Gem.loaded_specs['inspec'].version.to_s rescue Inspec::VERSION
```
Comment (clintoncwolfe) [inline test/integration/default/controls/audit_spec.rb:17]: Could you explain this change?

File: test/unit/matchers/matchers_test.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+require "rspec/expectations"
+require "matchers/matchers"
+
+require "pry"
+
+describe 'inspec matchers' do
+  describe 'cmp matcher' do
+    include RSpec::Matchers
+
+    def rspec_expect value, &block
```
Comment (clintoncwolfe) [inline test/unit/matchers/matchers_test.rb:17]: This probably needs in-code comments to explain what is going on. We shouldn't assume maintainers have any particular knowledge of RSpec.

File: test/unit/matchers/matchers_test.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      ::RSpec::Expectations::ExpectationTarget.for(value, block)
+    end
+
+    def assert_cmp exp, act
```
Comment (clintoncwolfe) [inline test/unit/matchers/matchers_test.rb]: I think `exp` means `expectation`, but I don't know what `act` refers to. Could you use more expressive variable names?

File: test/unit/plugin/v2/installer_test.rb [hunk 9]
```diff
@@ -51,18 +58,19 @@ def teardown
       Dir.glob(File.join(@config_dir_path, 'empty', '*')).each do |path|
         next if path.end_with? '.gitkeep'
         FileUtils.rm_rf(path)
-      end
+      end if @config_dir_path
     end
 
     # Clean up any activated gems
     Gem.loaded_specs.delete('inspec-test-fixture')
+    Gem.loaded_specs.delete('ordinal_array')
```
Comment (clintoncwolfe) [inline test/unit/plugin/v2/installer_test.rb:69]: I think I know what this fixes, but just for completeness, could you comment explaining what this is for?

File: test/unit/plugin/v2/installer_test.rb [hunk 10]
```diff
@@ -195,13 +209,18 @@ def test_install_a_pinned_gem_from_rubygems_org
   def test_install_a_gem_with_conflicting_depends_from_rubygems_org
     ENV['INSPEC_CONFIG_DIR'] = File.join(@config_dir_path, 'empty')
 
+    spec = Gem::Specification._all.find { |s| s.name == "rake" }
```
Comment (clintoncwolfe) [inline test/unit/plugin/v2/installer_test.rb:216]: Could you explain this?

**Review by clintoncwolfe — COMMENTED** [review]
This looks generally good; however on the sporadic failure fixes, there are some that I am not clear on - adding some comments on the PR indicating what fixes what would be very helpful. I also had a clarity concern or two.

Those aren't really blockers; if you add the comments, I can flip to Approve.

File: test/unit/plugin/v2/installer_test.rb (hunk 9 above)
Comment (zenspider) [reply test/unit/plugin/v2/installer_test.rb:69]: What are you looking for here beyond the comment above it? That some tests activate different dependencies and will wind up causing test ordering issues?

File: test/unit/matchers/matchers_test.rb (hunk 8 above)
Comment (zenspider) [reply test/unit/matchers/matchers_test.rb]: "actual". I use this all over (esp in the doco for minitest) so I make assumptions about it's "obviousness". What I really want is a better spoken language that has same-length antonyms (true vs false, local vs remote, etc).

**Comment by zenspider** [comment]
This diff is bigger than it needs to be because it was rebased over master and then my modernize_minitest branch. If we merge the latter, then this diff should clean up a fair amount.

**Comment by miah** [comment]
Since we merged #4023 you might have to rebase this branch against master to clean that diff up.

File: lib/matchers/matchers.rb (hunk 3 above)
Comment (zenspider) [reply lib/matchers/matchers.rb:249]: They're not. They're part of the general cleanup of stuff that I saw as I went through. The main reason this should go in is this:

```
2896 % ./cmp.rb 
Warming up --------------------------------------
         method+call   154.352k i/100ms
                send   248.925k i/100ms
Calculating -------------------------------------
         method+call      2.603M (± 1.4%) i/s -     13.120M in   5.041969s
                send      6.425M (± 3.1%) i/s -     32.111M in   5.003057s

Comparison:
                send:  6424944.9 i/s
         method+call:  2602626.5 i/s - 2.47x  slower
```

Especially considering that this is on the matcher side of things, and thus, hit by all users.

File: test/unit/matchers/matchers_test.rb (hunk 8 above)
Comment (zenspider) [reply test/unit/matchers/matchers_test.rb]: fixed

File: test/unit/matchers/matchers_test.rb (hunk 7 above)
Comment (zenspider) [reply test/unit/matchers/matchers_test.rb:17]: fixed

**Comment by miah** [comment]
Lets rebase this now that #4066 has landed.

**Review by miah — APPROVED** [review]

File: lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb (hunk 4 above)
Comment (zenspider) [reply lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb:572]: We don't require bundler to run the test suite. It runs fine without it with the exception of some of these test assumptions. 

Requiring bundler of our users and especially contributors makes total sense, but the developers on this project shouldn't be beholden to an unnecessary requirement like bundler, especially when it just slows us down.

```
4402 % time ntimes 10 bundle exec ~/.rbenv/versions/2.6.2/bin/ruby -Ilib:test -e 0

real	0m5.506s
user	0m4.504s
sys	0m0.883s
4403 % time ntimes 10 ~/.rbenv/versions/2.6.2/bin/ruby -Ilib:test -e 0

real	0m0.876s
user	0m0.694s
sys	0m0.146s
```

Some of these commits are effectively decoupling our tests from bundler.

File: test/integration/default/controls/audit_spec.rb (hunk 6 above)
Comment (zenspider) [reply test/integration/default/controls/audit_spec.rb:17]: The active code. That is using our VERSION constant instead of finding the same value indirectly all the way through rubygems and back. If the spec isn't loaded, then that will fail out. There's no reason that the spec needs to be loaded for this test to run or pass.

File: test/unit/plugin/v2/installer_test.rb (hunk 10 above)
Comment (zenspider) [reply test/unit/plugin/v2/installer_test.rb:216]: Decouples from bundler.

**Comment by miah** [comment]
This makes us go GREEN on Travis. I'm going to merge this to get us into a better place. @clintoncwolfe please review this still.
